### PR TITLE
Add TileLink Address Remapper driver and example on qemu-sifive-e31-customer platform

### PR DESCRIPTION
**Pull request by frankchang0125**
Add TileLink Address Remapper driver and example.

To build example:
`make PROGRAM=example-remapper TARGET=qemu-sifive-e31-customer CONFIGURATION=release software`

QEMU with address remapper support: [https://github.com/sifive/qemu/tree/sifive-e-customer-v6.0.0](https://github.com/sifive/qemu/tree/sifive-e-customer-v6.0.0)

To run example with QEMU:
`qemu-system-riscv32 -nographic -M sifive_e_customer -bios none -kernel <your path to example>/example-remapper.elf`

Example output:
```
Address Remapper version: 1
From region base address: 0x0
From region size: 0x100000000 bytes
To region base address: 0x0
To region size: 0x100000000 bytes
Maximum From entry region size: 0x100000 bytes

Remap entry - From[]: 0x20400377
Remap entry - To[]: 0x80000000
Remap rom_data to ram_data, remap size: 16 bytes.
rom_data address: 0x20400370
ram_data address: 0x80000000
Enable remap...
rom_data[0] = 5
rom_data[1] = 6
rom_data[2] = 7
rom_data[3] = 8
remappervalid[3] = 0x00020000

Disable remap...
rom_data[0] = 1
rom_data[1] = 2
rom_data[2] = 3
rom_data[3] = 4
remappervalid[3] = 0x00000000

Enable remap...
rom_data[0] = 5
rom_data[1] = 6
rom_data[2] = 7
rom_data[3] = 8
remappervalid[3] = 0x00020000

Flush all remaps...
rom_data[0] = 1
rom_data[1] = 2
rom_data[2] = 3
rom_data[3] = 4

remappervalid[0] = 0x00000000
remappervalid[1] = 0x00000000
remappervalid[2] = 0x00000000
remappervalid[3] = 0x00000000
remappervalid[4] = 0x00000000
remappervalid[5] = 0x00000000
remappervalid[6] = 0x00000000
End of example.
```

### CODE REVIEW

**Comment by paul-walmsley-sifive** [comment]
Hi @frankchang0125 , looks like this one is failing one of the FESDK CI checks.  Is this expected?

**Comment by nick-knight** [comment]
Hey @frankchang0125 , the reason your `printf`s aren't working is because you are attempting to use `%lx` to print `uint64_t` on RV32. The portable approach is to use `PRIx64` [[cppreference]](https://en.cppreference.com/w/c/types/integer), which will expand to `%llx` on RV32 ... and which isn't supported by Newlib-Nano, the default C standard library for FESDK, along with other C99 print-formatting functionality. I reported this over a year ago:

https://sifive.atlassian.net/browse/ESAT-237

and the Embedded Software team ultimately decided not to fix it: arguably it's a feature, not a bug, and anyway it's supposed to be addressed by PicoLibc.

Long story short, if you want to print 64-bit integers on RV32 using FESDK (with default C library), you're going to have to write your own `printf`.  Here's the workaround my team uses: 

https://github.com/sifive/perf-tools-sifive/blob/master/src/util.h#L60-L75

**Comment by frankchang0125** [comment]
> Hi @frankchang0125 , looks like this one is failing one of the FESDK CI checks. Is this expected?

@paul-walmsley-sifive As I've made some changes to **freedom-devicetree-tools**, I think I should update all the BSP files instead just the qemu-sifive-e31 which I was using. I'll fix it and push new commits to fix it.

However, it looks like the DTS naming is still not finalized yet: [link](https://github.com/sifive/federation/pull/13463#issuecomment-822082486). So I think this PR has to be hold on until the DTS is been finalized, as **freedom-devicetree-tools** and **freedom-metal driver** also have to make the related changes.

**Comment by frankchang0125** [comment]
> Hey @frankchang0125 , the reason your `printf`s aren't working is because you are attempting to use `%lx` to print `uint64_t` on ILP32. The portable approach is to use the `PRIx64` macro [[cppreference]](https://en.cppreference.com/w/c/types/integer), which expands to `"llx"` on ILP32 ...

Thanks Nick, I didn't realized I miss-typed `llx` to `lx`. I will make the fix in the following commits.

> and which --- along with other C99 print-formatting functionality --- isn't supported by Newlib-Nano, the default C standard library for FESDK. I reported this over a year ago:
> 
> https://sifive.atlassian.net/browse/ESAT-237
> 
> and the Embedded Software team ultimately decided not to fix it: arguably it's a feature of Newlib-Nano, not a bug in FESDK, and there are workarounds. Here's the workaround my team uses (our own `printf`):
> 
> https://github.com/sifive/perf-tools-sifive/blob/master/src/util.h#L60-L75

Thanks, I may try the approach your team is using.

> 
> Alternatively, you can ask FESDK to use a different C library that implements C99 `printf` functionality, like (vanilla) Newlib. I forget how to do this but I know there is a way. This does bloat the binaries substantially --- there's a reason Newlib-Nano sacrificed this functionality --- so you may get link-time errors on smaller cores that you didn't have before.

As French team is building our next-version FESDK ([freedom-boot](https://github.com/sifive/freedom-boot)?). Hope that they've considered the issue you raised.

**Comment by frankchang0125** [comment]
I've updated the commits to pass all CI tests.
Also current example follows the latest changes of [address remapper spec](https://github.com/sifive/arch-specs/pull/406) and [hardware designs](https://github.com/sifive/federation/commit/be0a53ac8292bb887154f5ac1d03385337017c23).
@e-puerto would you please help to review this PR? Thanks.

**Comment by frankchang0125** [comment]
Updated example, driver and freedom-devicetree-tools to include new introduced DTS property: `sifive,remapper-max-frm-width`.

File: bsp/qemu-sifive-e31mc-customer/qemu.cfg [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# qemu config file
+
+[machine]
+  type = "sifive_e"
+
```
Comment (e-puerto) [inline bsp/qemu-sifive-e31mc-customer/qemu.cfg:5]: I think it missing 
```
[smp-opts]
  cpus = "4"
```

File: bsp/qemu-sifive-e31-customer/README.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+- 4 hardware breakpoints
+- 16 local interrupts signal that can be connected to off core complex devices
+- Up to 127 PLIC interrupt signals that can be connected to off core complex devices, with 7 priority levels
+- GPIO memory with 16 interrupt lines
+- SPI memory with 1 interrupt line
+- Serial port with 1 interrupt line
+- 1 RGB LEDS
+- DTIM memory with 4M bytes
+
+This BSP matches the QEMU code in https://github.com/sifive/freedom-qemu
```
Comment (e-puerto) [inline bsp/qemu-sifive-e31-customer/README.md:16]: Could you specify the particular branch, or tag, or minimal version of qemu ?

File: bsp/qemu-sifive-e31-customer/settings.mk [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+RISCV_ARCH = rv32imac
+RISCV_ABI = ilp32
+RISCV_CMODEL = medlow
+RISCV_SERIES = None
+
+TARGET_TAGS = qemu
+TARGET_DHRY_ITERS = 20000000
+TARGET_CORE_ITERS = 5000
+TARGET_FREERTOS_WAIT_MS = 1000
+TARGET_INTR_WAIT_CYCLE  = 0
```
Comment (e-puerto) [inline bsp/qemu-sifive-e31-customer/settings.mk:13]: add newline ;)

File: bsp/qemu-sifive-e31mc-customer/README.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+- 4 hardware breakpoints
+- 16 local interrupts signal that can be connected to off core complex devices
+- Up to 127 PLIC interrupt signals that can be connected to off core complex devices, with 7 priority levels
+- GPIO memory with 16 interrupt lines
+- SPI memory with 1 interrupt line
+- Serial port with 1 interrupt line
+- 1 RGB LEDS
+- DTIM memory with 4M bytes
+
+This BSP matches the QEMU code in https://github.com/sifive/freedom-qemu
```
Comment (e-puerto) [inline bsp/qemu-sifive-e31mc-customer/README.md:16]: Could you specify the particular branch, or tag, or minimal version of qemu?

File: bsp/qemu-sifive-e31mc-customer/settings.mk [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+RISCV_ARCH = rv32imac
+RISCV_ABI = ilp32
+RISCV_CMODEL = medlow
+RISCV_SERIES = None
+
+TARGET_TAGS = qemu
+TARGET_DHRY_ITERS = 20000000
+TARGET_CORE_ITERS = 5000
+TARGET_FREERTOS_WAIT_MS = 1000
+TARGET_INTR_WAIT_CYCLE  = 0
```
Comment (e-puerto) [inline bsp/qemu-sifive-e31mc-customer/settings.mk:13]: add newline ;)

**Review by e-puerto — CHANGES_REQUESTED** [review]
some minor change ;)

File: bsp/qemu-sifive-e31-customer/settings.mk (hunk 3 above)
Comment (frankchang0125) [reply bsp/qemu-sifive-e31-customer/settings.mk]: _settings.mk_ is generated by _update-targets.sh_ and it doesn't have new line at the end of file.
I've made the fix, but the CI tests will fail. I will revert my change back.

File: bsp/qemu-sifive-e31mc-customer/settings.mk (hunk 5 above)
Comment (frankchang0125) [reply bsp/qemu-sifive-e31mc-customer/settings.mk]: Same as: https://github.com/sifive/freedom-e-sdk/pull/589#discussion_r644084325
